File: app/src/main/java/com/tsymiar/devidroid/utils/Pcm2Wav.java
package com.tsymiar.devidroid.utils;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * @author cokus
 * 将pcm转换成wav格式
 * 其实就是加入wav头
 */
class Pcm2Wav {

    public class WaveHeader {
        final char[] fileID = {'R', 'I', 'F', 'F'};

        int fileLength;

        char[] wavTag = {'W', 'A', 'V', 'E'};

        char[] FmtHdrID = {'f', 'm', 't', ' '};

        int FmtHdrLen;

        short FormatTag;

        short Channels;

        int SamplesPerSec;

        int AvgBytesPerSec;

        short BlockAlign;

        short BitsPerSample;

        char[] DataHdrID = {'d', 'a', 't', 'a'};

        int DataHdrLen;

        byte[] getHeader() throws IOException {
            ByteArrayOutputStream bos = new ByteArrayOutputStream();
            WriteChar(bos, fileID);
            WriteInt(bos, fileLength);
            WriteChar(bos, wavTag);
            WriteChar(bos, FmtHdrID);
            WriteInt(bos, FmtHdrLen);
            WriteShort(bos, FormatTag);
            WriteShort(bos, Channels);
            WriteInt(bos, SamplesPerSec);
            WriteInt(bos, AvgBytesPerSec);
            WriteShort(bos, BlockAlign);
            WriteShort(bos, BitsPerSample);
            WriteChar(bos, DataHdrID);
            WriteInt(bos, DataHdrLen);
            bos.flush();
            byte[] r = bos.toByteArray();
            bos.close();
            return r;
        }

        private void WriteShort(ByteArrayOutputStream bos, int s)
                throws IOException {
            byte[] mybyte = new byte[2];
            mybyte[1] = (byte) ((s << 16) >> 24);
            mybyte[0] = (byte) ((s << 24) >> 24);
            bos.write(mybyte);
        }

        private void WriteInt(ByteArrayOutputStream bos, int n) throws IOException {
            byte[] buf = new byte[4];
            buf[3] = (byte) (n >> 24);
            buf[2] = (byte) ((n << 8) >> 24);
            buf[1] = (byte) ((n << 16) >> 24);
            buf[0] = (byte) ((n << 24) >> 24);
            bos.write(buf);
        }

        private void WriteChar(ByteArrayOutputStream bos, char[] id) {
            for (char c : id) {
                bos.write(c);
            }
        }
    }

    void convertAudioFiles(String src, String target) throws Exception {
        FileInputStream fis = new FileInputStream(src);
        FileOutputStream fos = new FileOutputStream(target);


        byte[] buf = new byte[1024 * 1000];
        int size = fis.read(buf);
        int PCMSize = 0;
        while (size != -1) {
            PCMSize += size;
            size = fis.read(buf);
        }
        fis.close();

        WaveHeader header = new WaveHeader();
        header.fileLength = PCMSize + (44 - 8);
        header.FmtHdrLen = 16;
        header.BitsPerSample = 16;
        header.Channels = 1;
        header.FormatTag = 0x0001;
        header.SamplesPerSec = 16000;
        header.BlockAlign = (short) (header.Channels * header.BitsPerSample / 8);
        header.AvgBytesPerSec = header.BlockAlign * header.SamplesPerSec;
        header.DataHdrLen = PCMSize;

        byte[] h = header.getHeader();

        if (h.length != 44) {
            return;
        }
        //write header
        fos.write(h, 0, h.length);
        //write data stream
        fis = new FileInputStream(src);
        size = fis.read(buf);
        while (size != -1) {
            fos.write(buf, 0, size);
            size = fis.read(buf);
        }
        fis.close();
        fos.close();
    }
}

File: app/src/main/cpp/decode/Pcm2Wav.cpp
#include <cstdarg>
#include <cstring>
#include <cstdio>

#ifndef LOG_TAG
#define LOG_TAG "Pcm2Wav"
#endif

#include <Utils/logging.h>
#include <files/FileUtils.h>

const char wavTag[] = {'W', 'A', 'V', 'E'};
const char fileID[] = {'R', 'I', 'F', 'F'};
const char fmtHdrID[] = {'f', 'm', 't', ' '};
const char dataHdrID[] = {'d', 'a', 't', 'a'};

int writeShort(char *l, int s)
{
    char byte[2];
    byte[1] = (char) (s << 16 >> 24);
    byte[0] = (char) (s << 24 >> 24);
    memset(l, s, 2);
    return 2;
}

int writeInt(char *l, int n)
{
    char buf[4];
    buf[3] = (char) (n >> 24);
    buf[2] = (char) (n << 8 >> 24);
    buf[1] = (char) (n << 16 >> 24);
    buf[0] = (char) (n << 24 >> 24);
    memcpy(l, buf, 4);
    return 4;
}

int writeChars(char *l, const char id[], size_t s)
{
    memcpy(l, id, s);
    return s;
}

class WaveHeader {
public:
    struct Content {
        char WavTag[4];
        short FormatTag;

        char FileID[4];
        char FmtHdrID[4];
        char DataHdrID[4];

        int FileLen;
        int FmtHdrLen;
        int DataHdrLen;

        short Channels;
        short BlockAlign;
        short BitsPerSample;
        int SamplesPerSec;
        int AvgBytesPerSec;
    };

    static char *getHeader(Content content)
    {
        static char header[sizeof(WaveHeader)];
        memcpy(content.WavTag, wavTag, 4);
        memcpy(content.FileID, fileID, 4);
        memcpy(content.FmtHdrID, fmtHdrID, 4);
        memcpy(content.DataHdrID, dataHdrID, 4);

        int len = writeChars(header, content.FileID, 4);
        len = writeInt(header + len, content.FileLen);
        len = writeChars(header + len, content.WavTag, 4);
        len = writeChars(header + len, content.FmtHdrID, 4);
        len = writeInt(header + len, content.FmtHdrLen);
        len = writeShort(header + len, content.FormatTag);
        len = writeShort(header + len, content.Channels);
        len = writeInt(header + len, content.SamplesPerSec);
        len = writeInt(header + len, content.AvgBytesPerSec);
        len = writeShort(header + len, content.BlockAlign);
        len = writeShort(header + len, content.BitsPerSample);
        len = writeChars(header + len, content.DataHdrID, 4);
        writeInt(header + len, content.DataHdrLen);
        return header;
    }

};

int convertAudioFiles(const char *from, const char *target)
{
    size_t len = 1024 * 1000;
    char buf[len];
    int PCMSize = 0;
    WaveHeader::Content content{};
    content.FileLen = PCMSize + (44 - 8);
    content.FmtHdrLen = 16;
    content.BitsPerSample = 16;
    content.Channels = 1;
    content.FormatTag = 0x0001;
    content.SamplesPerSec = 16000;
    content.BlockAlign = (short) (content.Channels * content.BitsPerSample / 8);
    content.AvgBytesPerSec = content.BlockAlign * content.SamplesPerSec;
    content.DataHdrLen = PCMSize;

    //write header
    size_t size = 44;
    char *h = WaveHeader::getHeader(content);
    if (h == nullptr) {
        return -1;
    }
    memcpy(buf, h, size);

    //get data from stream
    FILE *fp = fopen(from, "r");
    if (fp == nullptr) {
        LOGE("can not load '%s' file!", from);
        return -2;
    }
    int c;
    while (EOF != (c = getc(fp))) {
        memset(buf + size, c, 1);
        size++;
    }
    fclose(fp);

    //write data stream
    FileUtils::MakeDirs(target);
    fp = fopen(target, "w");
    if (fp == nullptr) {
        LOGE("can not load '%s' file!", target);
        return -4;
    }
    if (fwrite(buf, 1, size, fp)) {
        LOGE("write target file '%s' failed.", target);
        return -3;
    }
    fclose(fp);
    return 0;
}
